File: backend/src/main/java/com/saif/portfolio/service/ProjectService.java
package com.saif.portfolio.service;

import java.util.List;

import org.springframework.data.domain.Page;

import com.saif.portfolio.dto.ProjectRequest;
import com.saif.portfolio.dto.ProjectResponse;
import com.saif.portfolio.dto.SimpleProjectResponse;

public interface ProjectService {

    List<ProjectResponse> getAllProjects();

    Page<SimpleProjectResponse> getAllSimpleProjectResponses(int page, int size);

    List<ProjectResponse> getFeaturedProjects();

    ProjectResponse getProjectById(int id);

    ProjectResponse getProjectBySlug(String slug);

    List<ProjectResponse> getProjectsByType(String type);

    ProjectResponse createProject(ProjectRequest request);

    ProjectResponse updateProject(Integer id, ProjectRequest request);

    ProjectResponse deleteProject(Integer id);

    ProjectResponse toggleLive(Integer id);

    ProjectResponse togglePublished(Integer id);

    ProjectResponse toggleFeatured(Integer id);
}

File: frontend/src/services/projectService.ts
import { apiClient } from "../lib/apiClient";

import { ApiResponse } from "../types/api";
import { ProjectRequest, ProjectResponse } from "../types/project";

class ProjectService {
  async getAllProjects(): Promise<ProjectResponse[]> {
    const response = await apiClient.get<ApiResponse<ProjectResponse[]>>("/projects");
    return response.data.data;
  }
  async getAllFeaturedProjects(): Promise<ProjectResponse[]> {
    const response = await apiClient.get<ApiResponse<ProjectResponse[]>>("/projects/featured");
    return response.data.data;
  }

  async getProjectById(id: number): Promise<ProjectResponse> {
    const response = await apiClient.get<ApiResponse<ProjectResponse>>(`/projects/${id}`);
    return response.data.data;
  }

  async getProjectBySlug(slug: string): Promise<ProjectResponse> {
    const response = await apiClient.get<ApiResponse<ProjectResponse>>(`/projects/slug/${slug}`);
    return response.data.data;
  }

  async getProjectsByType(type: string): Promise<ProjectResponse[]> {
    const response = await apiClient.get<ApiResponse<ProjectResponse[]>>(`/projects/type/${type}`);
    return response.data.data;
  }

  async createProject(project: ProjectRequest): Promise<ProjectResponse> {
    const response = await apiClient.post<ApiResponse<ProjectResponse>>("/projects", project);
    return response.data.data;
  }

  async updateProject(id: number, project: ProjectRequest): Promise<ProjectResponse> {
    const response = await apiClient.put<ApiResponse<ProjectResponse>>(`/projects/${id}`, project);
    return response.data.data;
  }

  async deleteProject(id: number): Promise<ProjectResponse> {
    const response = await apiClient.delete<ApiResponse<ProjectResponse>>(`/projects/${id}`);
    return response.data.data;
  }

  async toggleLive(id: number): Promise<ProjectResponse> {
    const response = await apiClient.patch<ApiResponse<ProjectResponse>>(`/projects/${id}/live`);
    return response.data.data;
  }

  async togglePublished(id: number): Promise<ProjectResponse> {
    const response = await apiClient.patch<ApiResponse<ProjectResponse>>(`/projects/${id}/published`);
    return response.data.data;
  }

  async toggleFeatured(id: number): Promise<ProjectResponse> {
    const response = await apiClient.patch<ApiResponse<ProjectResponse>>(`/projects/${id}/featured`);
    return response.data.data;
  }
}

const projectService = new ProjectService();
export default projectService;